data_safe_haven/pulumi/components/sre_research_desktop.py

```diff
@@ -53,19 +53,20 @@ def __init__(
             virtual_network_resource_group
         ).apply(get_name_from_rg)
         self.vm_ip_addresses = Output.all(subnet_research_desktops, vm_details).apply(
-            lambda args: self.get_ip_addresses(
-                cast(network.GetSubnetResult, args[0]), len(cast(List[str], args[1]))
-            )
+            lambda args: self.get_ip_addresses(subnet=args[0], vm_details=args[1])
         )
         vm_lists = Output.from_input(vm_details).apply(
             lambda d: [(name, details["sku"]) for name, details in d.items()]
         )
         self.vm_names = vm_lists.apply(lambda l: [t[0] for t in l])
         self.vm_sizes = vm_lists.apply(lambda l: [t[1] for t in l])
 
-    def get_ip_addresses(
-        self, subnet: network.GetSubnetResult, number: int
-    ) -> List[str]:
+    def get_ip_addresses(self, subnet: Any, vm_details: Any) -> List[str]:
+        assert isinstance(subnet, network.GetSubnetResult)
+        assert isinstance(List[str], vm_details)
```

Comment:
Looks good but I think `assert` isn't strong enough here. Better to explicitly use `if not isinstance(): raise` since `assert` can be optimised out.

Reply:
:+1: Looking at `data_safe_haven/exceptions/__init__.py`, would `DataSafeHavenPulumiException` be suitable here?